[SYCL] AST support for SYCL kernel entry point functions.

tahonermann opened the pull request:
A SYCL kernel entry point function is a non-member function or a static member function declared with the `sycl_kernel_entry_point` attribute. Such functions define a pattern for an offload kernel entry point function to be generated to enable execution of a SYCL kernel on a device. A SYCL library implementation orchestrates the invocation of these functions with corresponding SYCL kernel arguments in response to calls to SYCL kernel invocation functions specified by the SYCL 2020 specification.

The offload kernel entry point function (sometimes referred to as the SYCL kernel caller function) is generated from the SYCL kernel entry point function by a transformation of the function parameters followed by a transformation of the function body to replace references to the original parameters with references to the transformed ones. Exactly how parameters are transformed will be explained in a future change that implements non-trivial transformations. For now, it suffices to state that a given parameter of the SYCL kernel entry point function may be transformed to multiple parameters of the offload kernel entry point as needed to satisfy offload kernel argument passing requirements. Parameters that are decomposed in this way are reconstituted as local variables in the body of the generated offload kernel entry point function.

For example, given the following SYCL kernel entry point function definition:
```
template<typename KernelNameType, typename KernelType>
[[clang::sycl_kernel_entry_point(KernelNameType)]]
void sycl_kernel_entry_point(KernelType kernel) {
  kernel();
}
```

and the following call:
```
struct Kernel {
  int dm1;
  int dm2;
  void operator()() const;
};
Kernel k;
sycl_kernel_entry_point<class kernel_name>(k);
```

the corresponding offload kernel entry point function that is generated might look as follows (assuming `Kernel` is a type that requires decomposition):
```
void offload_kernel_entry_point_for_kernel_name(int dm1, int dm2) {
  Kernel kernel{dm1, dm2};
  kernel();
}
```

Other details of the generated offload kernel entry point function, such as its name and calling convention, are implementation details that need not be reflected in the AST and may differ across target devices. For that reason, only the transformation described above is represented in the AST; other details will be filled in during code generation.

These transformations are represented using new AST nodes introduced with this change. `OutlinedFunctionDecl` holds a sequence of `ImplicitParamDecl` nodes and a sequence of statement nodes that correspond to the transformed parameters and function body. `SYCLKernelCallStmt` wraps the original function body and associates it with an `OutlinedFunctionDecl` instance. For the example above, the AST generated for the `sycl_kernel_entry_point<kernel_name>` specialization would look as follows:
```
FunctionDecl 'sycl_kernel_entry_point<kernel_name>(Kernel)'
  TemplateArgument type 'kernel_name'
  TemplateArgument type 'Kernel'
  ParmVarDecl kernel 'Kernel'
  SYCLKernelCallStmt
    CompoundStmt
      <original statements>
    OutlinedFunctionDecl
      ImplicitParamDecl 'dm1' 'int'
      ImplicitParamDecl 'dm2' 'int'
      CompoundStmt
        VarDecl 'kernel' 'Kernel'
          <initialization of 'kernel' with 'dm1' and 'dm2'>
        <transformed statements with redirected references of 'kernel'>
```

Any ODR-use of the SYCL kernel entry point function will (with future changes) suffice for the offload kernel entry point to be emitted. An actual call to the SYCL kernel entry point function will result in a call to the function. However, evaluation of a `SYCLKernelCallStmt` statement is a no-op, so such calls will have no effect other than to trigger emission of the offload kernel entry point.

Additionally, as a related change inspired by code review feedback, these changes disallow use of the `sycl_kernel_entry_point` attribute with functions defined with a _function-try-block_. The SYCL 2020 specification prohibits the use of C++ exceptions in device functions. Even if exceptions were not prohibited, it is unclear what the semantics would be for an exception that escapes the SYCL kernel entry point function; the boundary between host and device code could be an implicit noexcept boundary that results in program termination if violated, or the exception could perhaps be propagated to host code via the SYCL library. Pending support for C++ exceptions in device code and clear semantics for handling them at the host-device boundary, this change makes use of the `sycl_kernel_entry_point` attribute with a function defined with a _function-try-block_ an error.

tahonermann committed:
[SYCL] AST support for SYCL kernel entry point functions.

A SYCL kernel entry point function is a non-member function or a static member
function declared with the `sycl_kernel_entry_point` attribute. Such functions
define a pattern for an offload kernel entry point function to be generated to
enable execution of a SYCL kernel on a device. A SYCL library implementation
orchestrates the invocation of these functions with corresponding SYCL kernel
arguments in response to calls to SYCL kernel invocation functions specified
by the SYCL 2020 specification.

The offload kernel entry point function (sometimes referred to as the SYCL
kernel caller function) is generated from the SYCL kernel entry point function
by a transformation of the function parameters followed by a transformation of
the function body to replace references to the original parameters with
references to the transformed ones. Exactly how parameters are transformed will
be explained in a future change that implements non-trivial transformations.
For now, it suffices to state that a given parameter of the SYCL kernel entry
point function may be transformed to multiple parameters of the offload kernel
entry point as needed to satisfy offload kernel argument passing requirements.
Parameters that are decomposed in this way are reconstituted as local variables
in the body of the generated offload kernel entry point function.

For example, given the following SYCL kernel entry point function definition:

  template<typename KernelNameType, typename KernelType>
  [[clang::sycl_kernel_entry_point(KernelNameType)]]
  void sycl_kernel_entry_point(KernelType kernel) {
    kernel();
  }

and the following call:

  struct Kernel {
    int dm1;
    int dm2;
    void operator()() const;
  };
  Kernel k;
  sycl_kernel_entry_point<class kernel_name>(k);

the corresponding offload kernel entry point function that is generated might
look as follows (assuming `Kernel` is a type that requires decomposition):

  void offload_kernel_entry_point_for_kernel_name(int dm1, int dm2) {
    Kernel kernel{dm1, dm2};
    kernel();
  }

Other details of the generated offload kernel entry point function, such as
its name and callng convention, are implementation details that need not be
reflected in the AST and may differ across target devices. For that reason,
only the transformation described above is represented in the AST; other
details will be filled in during code generation.

These transformations are represented using new AST nodes introduced with this
change. `OutlinedFunctionDecl` holds a sequence of `ImplicitParamDecl` nodes
and a sequence of statement nodes that correspond to the transformed
parameters and function body. `SYCLKernelCallStmt` wraps the original function
body and associates it with an `OutlinedFunctionDecl` instance. For the example
above, the AST generated for the `sycl_kernel_entry_point<kernel_name>`
specialization would look as follows:

  FunctionDecl 'sycl_kernel_entry_point<kernel_name>(Kernel)'
    TemplateArgument type 'kernel_name'
    TemplateArgument type 'Kernel'
    ParmVarDecl kernel 'Kernel'
    SYCLKernelCallStmt
      CompoundStmt
        <original statements>
      OutlinedFunctionDecl
        ImplicitParamDecl 'dm1' 'int'
        ImplicitParamDecl 'dm2' 'int'
        CompoundStmt
          VarDecl 'kernel' 'Kernel'
            <initialization of 'kernel' with 'dm1' and 'dm2'>
          <transformed statements with redirected references of 'kernel'>

Any ODR-use of the SYCL kernel entry point function will (with future changes)
suffice for the offload kernel entry point to be emitted. An actual call to
the SYCL kernel entry point function will result in a call to the function.
However, evaluation of a `SYCLKernelCallStmt` statement is a no-op, so such
calls will have no effect other than to trigger emission of the offload kernel
entry point.

tahonermann committed:
Squash; address the non-dumb clang-format complaints.

erichkeane commented on clang/include/clang/AST/Decl.h:
```diff
@@ -4678,6 +4678,95 @@ class BlockDecl : public Decl, public DeclContext {
+      public DeclContext,
+      private llvm::TrailingObjects<OutlinedFunctionDecl, ImplicitParamDecl *> {
+protected:
+  size_t numTrailingObjects(OverloadToken<ImplicitParamDecl>) {
```
This function isn't required/needed/used when there is only 1 trailing object type.

tahonermann replied:
Unneeded `numTrailingObjects()` member removed. Done.

erichkeane commented on clang/include/clang/AST/Decl.h:
```diff
@@ -4678,6 +4678,95 @@ class BlockDecl : public Decl, public DeclContext {
+    assert(i < NumParams);
+    return getParams()[i];
+  }
+  void setParam(unsigned i, ImplicitParamDecl *P) {
```
Does these have to be public?  These 'setters' should be immutable if at all possible (obviously besides serialization, but you're a friend)..

erichkeane commented on clang/include/clang/AST/Decl.h:
```diff
@@ -4678,6 +4678,95 @@ class BlockDecl : public Decl, public DeclContext {
+  using param_range = llvm::iterator_range<param_iterator>;
+
+  /// Retrieve an iterator pointing to the first parameter decl.
+  param_iterator param_begin() const { return getParams(); }
```
```suggestion
  param_iterator parameters_begin() const { return getParams(); }
```

Typically we name the 'range' function and the begin/end functions the same when sensible.  And I think it is sensible here? 

Also, please implement them in terms of each-other.

tahonermann replied:
Done, but without the rename as explained.

erichkeane commented on clang/include/clang/AST/Decl.h:
```diff
@@ -4678,6 +4678,95 @@ class BlockDecl : public Decl, public DeclContext {
+  }
+
+  using param_iterator = ImplicitParamDecl *const *;
+  using param_range = llvm::iterator_range<param_iterator>;
```
Unused?  

We should probably just use this iterator_range instead of ArrayRef in `parameters`.

tahonermann replied:
Done.

erichkeane commented on clang/include/clang/AST/StmtSYCL.h:
```diff
@@ -0,0 +1,92 @@
+  SYCLKernelCallStmt(EmptyShell Empty) : Stmt(SYCLKernelCallStmtClass, Empty) {}
+
+  /// Retrieve the model statement.
+  Stmt *getOriginalStmt() { return OriginalStmt; }
```
Can this be a `CompoundStmt` instead of `Stmt`? As that is what the description says it should always be, right?

erichkeane replied:
Ah, wait, `CXXTryStmt`.

tahonermann replied:
Done.

erichkeane replied:
That seems like a sensible resolution.

erichkeane commented on clang/lib/Serialization/ASTReaderDecl.cpp:
```diff
@@ -451,9 +452,9 @@ class ASTDeclReader : public DeclVisitor<ASTDeclReader, void> {
   void VisitOMPDeclareMapperDecl(OMPDeclareMapperDecl *D);
   void VisitOMPRequiresDecl(OMPRequiresDecl *D);
   void VisitOMPCapturedExprDecl(OMPCapturedExprDecl *D);
-  };
```
Unrelated changes.

erichkeane commented on clang/lib/Serialization/ASTReaderDecl.cpp:
```diff
@@ -451,9 +452,9 @@ class ASTDeclReader : public DeclVisitor<ASTDeclReader, void> {
-  };
+};
 
-  } // namespace clang
```
Same

erichkeane commented on clang/lib/Serialization/ASTReaderDecl.cpp:
```diff
@@ -1793,6 +1794,16 @@ void ASTDeclReader::VisitBlockDecl(BlockDecl *BD) {
+  // NumParams is deserialized by OutlinedFunctionDecl::CreateDeserialized().
+  VisitDecl(D);
+  D->setNothrow(Record.readInt() != 0);
+  for (unsigned I = 0; I < D->NumParams; ++I) {
```
No curleys for 1 liners.

erichkeane commented on clang/lib/Serialization/ASTWriterDecl.cpp:
```diff
@@ -1376,6 +1377,16 @@ void ASTDeclWriter::VisitBlockDecl(BlockDecl *D) {
+void ASTDeclWriter::VisitOutlinedFunctionDecl(OutlinedFunctionDecl *D) {
+  Record.push_back(D->getNumParams());
+  VisitDecl(D);
+  Record.push_back(D->isNothrow() ? 1 : 0);
```
Slight preference to move this 'down' and put the params and the num-params next to eachother.

tahonermann replied:
Done.

tahonermann committed:
Addressed the first round of code review comments from Erich Keane.

tahonermann committed:
Disallow applying the sycl_kernel_entry_point attribute to functions with a function-try-block.

The SYCL 2020 specification prohibits the use of C++ exceptions in device
functions. Even if exceptions were not prohibited, it is unclear what the
semantics would be for an exception that escapes the SYCL kernel entry point
function; the boundary between host and device code could be an implicit
noexcept boundary that results in program termination if violated, or the
exception could perhaps be propagated to host code via the SYCL library.
The semantics of function-try-blocks is that an exception raised in the
function body that escapes to a function-try-block handler is implicitly
rethrown. Pending support for C++ exceptions in device code and clear
semantics for handling them at the host-device boundary, this change makes
use of the sycl_kernel_entry_point attribute with a function defined with
a function-try-block an error.

tahonermann commented:
@erichkeane, thank you for the prior code review. I believe I've addressed all the concerns you raised. There are two conversations I left unresolved pending any further response from you.

erichkeane commented on clang/include/clang/AST/StmtSYCL.h:
```diff
@@ -0,0 +1,94 @@
+
+} // end namespace clang
+
+#endif
```
Nit: this needs the comment on it i think that shows the macro .
```suggestion
#endif //LLVM_CLANG_AST_STMTSYCL_H
```

erichkeane commented on clang/include/clang/Basic/AttrDocs.td:
```diff
@@ -480,6 +480,7 @@ following requirements.
 * Is not a C variadic function.
 * Is not a coroutine.
 * Is not defined as deleted or as defaulted.
+* Is not defined with a function try block.
```
Did we submit an issue against the SYCL spec here?  Seems sensible to do (and could we do something like:)
```suggestion
* Is not defined with a function try block. See SYCL-CWG1234
```

erichkeane commented on clang/lib/Sema/JumpDiagnostics.cpp:
```diff
@@ -18,6 +18,7 @@
 #include "clang/AST/StmtObjC.h"
 #include "clang/AST/StmtOpenACC.h"
 #include "clang/AST/StmtOpenMP.h"
+#include "clang/AST/StmtSYCL.h"
```
Why is this here?

tahonermann replied:
Hmm, looks like no reason at all. I'm guessing it was once needed for some X-Macro usage prior to some rebase. I'll remove it.

erichkeane reviewed (commented):
I only saw 1 unresolved, not 2? But I resolved it.  Can you point out the other?

erichkeane reviewed (approved):
3 comments are all 'nits', so once those conversations are resolved, this LGTM.  Please resolve those 3 before merging.

tahonermann committed:
Addressed the second round of code review comments from Erich Keane.

tahonermann reviewed (commented):
> I only saw 1 unresolved, not 2? But I resolved it. Can you point out the other?

The other one was [this one](https://github.com/llvm/llvm-project/pull/122379#discussion_r1910706213); You gave it a thumbs up, so I went ahead and resolved it now.

erichkeane commented:
No idea why I can't respond to the individual threads but...

>The coding standards don't specify to add a comment there and a brief survey of other headers in this directory indicates considerable inconsistency (72 out of 123 .h files have them). I guess that is over 50% so I'll add it.

It kinda does?  Clang-format is supposed to be adding those and coding standard is that we're supposed to be clang-format clean.  That said, I've got no idea why it didn't try to add it here, I've definitely seen it do so before.

erichkeane reviewed (approved):
make format happy, then LGTM.

tahonermann committed:
clang-format is a harsh mistress.

tahonermann commented:
> No idea why I can't respond to the individual threads but...

@erichkeane, in order to reduce notifications, I replied to your comments as a review. You should generally be able to reply to them individually or as part of a batched review from the "Files changed" tab, but GitHub doesn't permit such replies from the "Conversation" tab (which is weird and seems dumb). However, if the comments are considered "outdated" because of new commits that changed the code the thread was anchored too, then such comments appear to become orphaned; they still appear in the "Conversation" tab, but not in the "Filed changed" tab and therefore can't be replied to. GitHub is a less than perfect tool, IMHO of course.

> make format happy, then LGTM.

Thank you. I addressed the non-dumb clang-format complaints. I'll merge once @bader is happy with the changes.

bader commented on clang/lib/Sema/SemaSYCL.cpp:
```diff
@@ -362,3 +364,93 @@ void SemaSYCL::CheckSYCLEntryPointFunctionDecl(FunctionDecl *FD) {
+// The body of a function declared with the [[sycl_kernel_entry_point]]
+// attribute is cloned and transformed to substitute references to the original
+// function parameters with references to replacement variables that stand in
+// for SYCL kernel parameters or local variables that reconstitute a decomposed
+// SYCL kernel argument.
```
I just want to confirm that we are not limited to adding new transforms (even ones that are not currently used in our downstream repository).
We may need new transforms to implement more efficient host <-> device SYCL kernel argument data transfer. We should be able to "compound" SYCL kernel arguments (i.e. the opposite of decomposing) if we find that more efficient. Right?

tahonermann replied:
@bader, we don't have an upstream design/implementation yet for handling decomposition and reconstitution of kernel arguments. My expectation is that we'll do something similar to what DPC++ does; there will be default rules for types generally with support for some types to be tagged as special SYCL types (types that have to be passed as independent kernel arguments) that force decomposition for some types that otherwise wouldn't require it. We'll likewise need to add support for types that have custom decomposition behavior similar to what DPC++ provides. I'm not aware of prior discussion regarding implicitly aggregating multiple kernel arguments into a single bundle argument. I'm curious how often doing so would actually be advantageous.

I am uneasy with the current DPC++ implementation since it requires the SYCL library and the front end to separately implement the decomposition rules. That seems like a recipe for bugs and means the library and front end have to be changed in lock step. I've been thinking about ways we can have the front end expose an interface to the library to perform or guide the decomposition so that the library can remain mostly agnostic of such details, but I don't have a design to share yet. The more we can isolate the library from such details, the more options we have for making changes.

bader reviewed (approved):
@tahonermann thanks a lot for the work. I really appreciate the detailed description and comments!

@erichkeane did a great job with the code review, so I have nothing to add.

I just have one small question that does not affect the merge.
